src/BalanceTx.purs

```diff
@@ -549,137 +551,142 @@ logTx msg utxos (Transaction { body: body'@(TxBody body) }) =
+        except $
+          worker unattachedTx' fees' (fees' - fees)
+  where
+  worker
+    :: UnattachedUnbalancedTx
+    -> BigInt
+    -> BigInt
+    -> Either ReturnAdaChangeError UnattachedUnbalancedTx
+  worker unattachedTx' fees feesDelta
+    | feesDelta <= zero = Right $
+        unattachedTxSetFees unattachedTx' fees
+    | otherwise =
+        let
+          txOutputs :: Array TransactionOutput
+          txOutputs = unattachedTx' ^. _body' <<< _outputs
 
-      txOutputs :: Array TransactionOutput
-      txOutputs = txBody.outputs
+          returnAda :: BigInt
+          returnAda = fromMaybe zero $
+            Array.head txOutputs <#> \(TransactionOutput rec) ->
+              (valueToCoin' rec.amount) - feesDelta
 
-      inputValue :: Value
-      inputValue = getInputValue utxos (wrap txBody)
+          utxoCost :: BigInt
+          utxoCost = getLovelace protocolParamUTxOCostPerWord
 
-      inputAda :: BigInt
-      inputAda = getLovelace $ valueToCoin inputValue
+          changeMinUtxo :: BigInt
+          changeMinUtxo = adaOnlyWords * utxoCost
+        in
+          case returnAda >= changeMinUtxo of
+            true -> do
+              newOutputs <- updateChangeTxOutputValue returnAda txOutputs
+              pure $
+                unattachedTx' # _body' %~ \(TxBody txBody) ->
+                  wrap txBody { outputs = newOutputs, fee = wrap fees }
+            false ->
+              Left $
+                ReturnAdaChangeError
+                  "returnAda does not cover min utxo requirement for \
+                  \single Ada-only output."
+
+  unattachedTxSetFees
+    :: UnattachedUnbalancedTx -> BigInt -> UnattachedUnbalancedTx
+  unattachedTxSetFees unattachedTx' fees =
+    unattachedTx' #
+      _body' <<< _fee .~ wrap fees
+
+  updateChangeTxOutputValue
+    :: BigInt
+    -> Array TransactionOutput
+    -> Either ReturnAdaChangeError (Array TransactionOutput)
+  updateChangeTxOutputValue returnAda =
+    note (ReturnAdaChangeError "Couldn't modify utxo to return change.")
+      <<< modifyAt zero
+        \(TransactionOutput rec) -> TransactionOutput
+          rec { amount = lovelaceValueOf returnAda }
 
-      -- FIX ME, ignore mint value?
-      outputValue :: Value
-      outputValue = Array.foldMap getAmount txOutputs
+returnAdaChange
+  :: Address
+  -> Utxo
+  -> UnattachedUnbalancedTx /\ BigInt
+  -> Either ReturnAdaChangeError
+       (UnattachedUnbalancedTx /\ { recalculateFees :: Boolean })
+returnAdaChange changeAddr utxos (unattachedTx /\ fees) =
+  let
+    TxBody txBody = unattachedTx ^. _body'
 
-      outputAda :: BigInt
-      outputAda = getLovelace $ valueToCoin outputValue
+    txOutputs :: Array TransactionOutput
+    txOutputs = txBody.outputs
+
+    inputValue :: Value
+    inputValue = getInputValue utxos (wrap txBody)
+
+    inputAda :: BigInt
+    inputAda = getLovelace $ valueToCoin inputValue
+
+    -- FIX ME, ignore mint value?
```

Comment:
Please open an issue for this with more details and include the URL here

Reply:
@ngua Honestly, I can't think of any particular reason why we should ignore mint value there, so I believe we can just remove this FIXME

Reply:
I asked Viet

Reply:
@klntsky Thanks

Reply:
Viet said we can really remove this FIXME 

> it seems redundant given we've only using it for outputAda and you can't mint Ada anyway.